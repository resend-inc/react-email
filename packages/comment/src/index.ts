export * from "./comment";

export * from "./mso";

export * from "./non-mso";

export * from "./mso-ghost";

export * from "./non-mso-ghost";
