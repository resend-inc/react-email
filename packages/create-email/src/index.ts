#!/usr/bin/env node

import { program } from "commander";
import logSymbols from "log-symbols";
import ora from "ora";
import pkg from "../package.json";
import { askUserThroughPrompts } from "./ask-user-through-prompts";
import { createStarter } from "./create-starter";
import { tree } from "./tree";

program
  .name("create-email")
  .version((pkg as { version: string }).version)
  .description("The easiest way to get started with React Email")
  .action(async () => {
    const { isTailwindEnabled, isTypescriptEnabled, absoluteProjectPath } =
      await askUserThroughPrompts();

    const spinner = ora("Preparing files...\n").start();

    await createStarter({
      absoluteProjectPath,
      enableTypeScript: isTypescriptEnabled,
      enableTailwindCSS: isTailwindEnabled,
    });

    console.log(await tree(absoluteProjectPath, 4));

    spinner.stopAndPersist({
      symbol: logSymbols.success,
      text: "React Email Starter files ready",
    });
  })
  .parse(process.argv);
