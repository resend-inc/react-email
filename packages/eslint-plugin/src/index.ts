import type { TSESLint } from "@typescript-eslint/utils";
import * as Rules from "./rules";
import { getAllSupportEntrries } from "./data/get-all-support-entries";
import { fromCamelToKebabCase } from "./casing/from-camel-to-kebab-case";
import { separateEntriesByCategory } from "./data/separate-entries-by-category";

const caniemailSupportEntries = getAllSupportEntrries();

const entriesByCategory = separateEntriesByCategory(caniemailSupportEntries);

const rules = Object.fromEntries(
  Object.entries(Rules).map(([ruleCamelCasedName, ruleFactory]) => [
    fromCamelToKebabCase(ruleCamelCasedName),
    ruleFactory(entriesByCategory),
  ]),
);

const plugin: TSESLint.Linter.Plugin = {
  configs: {
    recommended: {
      rules: {
        "react/jsx-key": "off",

        // rules for gmail
        "@react-email/no-unsupported-on-gmail-desktop-webmail": "error",
        "@react-email/no-partially-supported-on-gmail-desktop-webmail": "warn",
        "@react-email/no-unsupported-on-gmail-ios": "error",
        "@react-email/no-partially-supported-on-gmail-ios": "warn",
        "@react-email/no-unsupported-on-gmail-android": "error",
        "@react-email/no-partially-supported-on-gmail-android": "warn",
        "@react-email/no-unsupported-on-gmail-mobile-webmail": "error",
        "@react-email/no-partially-supported-on-gmail-mobile-webmail": "warn",

        // rules for apple mail
        "@react-email/no-unsupported-on-apple-mail-macos": "error",
        "@react-email/no-partially-supported-on-apple-mail-macos": "warn",
        "@react-email/no-unsupported-on-apple-mail-ios": "error",
        "@react-email/no-partially-supported-on-apple-mail-ios": "warn",

        // rules for outlook
        "@react-email/no-unsupported-on-outlook-windows": "error",
        "@react-email/no-partially-supported-on-outlook-windows": "warn",
        "@react-email/no-unsupported-on-outlook-windows-mail": "error",
        "@react-email/no-partially-supported-on-outlook-windows-mail": "warn",
        "@react-email/no-unsupported-on-outlook-macos": "error",
        "@react-email/no-partially-supported-on-outlook-macos": "warn",
        "@react-email/no-unsupported-on-outlook-outlook-com": "error",
        "@react-email/no-partially-supported-on-outlook-outlook-com": "warn",
        "@react-email/no-unsupported-on-outlook-ios": "error",
        "@react-email/no-partially-supported-on-outlook-ios": "warn",
        "@react-email/no-unsupported-on-outlook-android": "error",
        "@react-email/no-partially-supported-on-outlook-android": "warn",

        // rules for yahoo
        "@react-email/no-unsupported-on-yahoo-desktop-webmail": "error",
        "@react-email/no-partially-supported-on-yahoo-desktop-webmail": "warn",
        "@react-email/no-unsupported-on-yahoo-ios": "error",
        "@react-email/no-partially-supported-on-yahoo-ios": "warn",
        "@react-email/no-unsupported-on-yahoo-android": "error",
        "@react-email/no-partially-supported-on-yahoo-android": "warn",

        // rules for hey
        "@react-email/no-unsupported-on-hey-desktop-webmail": "error",
        "@react-email/no-partially-supported-on-hey-desktop-webmail": "warn",
      },
    },
  },
  rules,
};

export default plugin;
