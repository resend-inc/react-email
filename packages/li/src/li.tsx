import * as React from 'react';

type LiElement = React.ElementRef<'li'>;
type RootProps = React.ComponentPropsWithoutRef<'li'>;

export interface LiProps extends RootProps {}

export const Li = React.forwardRef<LiElement, Readonly<LiProps>>(
  ({ children, style, ...props }, forwardedRef) => (
    <li
      ref={forwardedRef}
      style={{
        ...style,
      }}
      {...props}
    >
      {children}
    </li>
  ),
);

Li.displayName = 'Li';
