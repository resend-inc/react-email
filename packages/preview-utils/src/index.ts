export * from './render-email-by-path';
export * from './get-email-component';
export * from './utils/email-template';
export * from './utils/rendering-utilities-exporter';

