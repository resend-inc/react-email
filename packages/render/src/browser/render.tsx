import { convert } from "html-to-text";
import { Suspense } from "react";
import { renderToReadableStream } from "react-dom/server";
import { Options } from "../shared/options";
import { plainTextSelectors } from "../shared/plain-text-selectors";
import { pretty } from "../shared/pretty";

async function streamToString(
  stream: ReadableStream<Uint8Array>,
): Promise<string> {
  const reader = stream.getReader();
  const decoder = new TextDecoder("utf-8");
  let result = "";
  let done: boolean | undefined;

  while (!done) {
    // eslint-disable-next-line no-await-in-loop
    const { value, done: doneReading } = await reader.read();
    done = doneReading;
    if (value) {
      result += decoder.decode(value, { stream: !done });
    }
  }

  return result;
}

export async function render(
  element: React.ReactElement,
  options?: Options,
): Promise<string> {
  const suspendedElement = <Suspense>{element}</Suspense>;
<<<<<<< HEAD
  const reactDOMServer = await import("react-dom/server");

  let html!: string;
  if (Object.hasOwn(reactDOMServer, "renderToReadableStream")) {
    html = await readStream(
      await reactDOMServer.renderToReadableStream(suspendedElement),
    );
  } else {
    await new Promise<void>((resolve, reject) => {
      const stream = reactDOMServer.renderToPipeableStream(suspendedElement, {
        async onAllReady() {
          html = await readStream(stream);
          resolve();
        },
        onError(error) {
          reject(error as Error);
        },
      });
    });
  }
=======
  const readableStream = await renderToReadableStream(suspendedElement);
  const html = await streamToString(readableStream);
>>>>>>> 447c7780 (refactor: `render` package)

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, "")}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
}
