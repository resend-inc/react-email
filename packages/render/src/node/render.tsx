import { Writable } from "node:stream";
import { renderToPipeableStream } from "react-dom/server";
import { Suspense } from "react";
import { convert } from "html-to-text";
import { Options } from "../shared/options";
import { pretty } from "../shared/pretty";
import { plainTextSelectors } from "../shared/plain-text-selectors";

export async function render(
  element: React.ReactElement,
  options?: Options,
): Promise<string> {
  const suspendedElement = <Suspense>{element}</Suspense>;
  const html = await new Promise<string>((resolve, reject) => {
    let htmlChunksString = "";

    const writable = new Writable({
      write(chunk: Buffer, _encoding, callback) {
        htmlChunksString += chunk.toString();
        callback();
      },
    });

    writable.on("finish", () => {
      resolve(htmlChunksString);
    });

    writable.on("error", (err) => {
      reject(err);
    });

    const stream = renderToPipeableStream(suspendedElement, {
      onAllReady() {
        stream.pipe(writable);
      },
      onError(error) {
        reject(error as Error);
      },
    });
  });

  if (options?.plainText) {
    return convert(html, {
      selectors: plainTextSelectors,
      ...options.htmlToTextOptions,
    });
  }

  const doctype =
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">';

  const document = `${doctype}${html.replace(/<!DOCTYPE.*?>/, "")}`;

  if (options?.pretty) {
    return pretty(document);
  }

  return document;
}
