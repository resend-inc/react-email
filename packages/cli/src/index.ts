import { parseArgs } from 'node:util';

import debugConfig from 'debug';
import importLocal from 'import-local';

import pkg from '../package.json';

import { command as build } from './commands/build';
import { command as help } from './commands/help';
import { command as preview } from './commands/preview';
import { CommandFn } from './commands/types';

const commands: Record<string, CommandFn> = { build, help, preview };
const debug = debugConfig('@jsx-email/cli');
const { log } = console;

const run = () => {
  const argv = parseArgs({ allowPositionals: true, args: process.argv.slice(2), strict: false });
  const [commandName] = argv.positionals;
  let command = commands[commandName];

  if (argv.values.version) {
    log(`${pkg.name} v${pkg.version}\n`);
  }

  if (!command) command = help;

  command(argv.values, argv.positionals || []);
};

if (importLocal(__filename)) {
  debug('Using local install of webpack-command');
} else {
  run();
}
