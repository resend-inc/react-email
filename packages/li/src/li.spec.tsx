import { Ol } from '../../ol/src/index';
import { Ul } from '../../ul/src/index';
import { Li } from './index';

import { render } from '@react-email/render';

describe('render', () => {
  beforeEach(() => {
    jest.restoreAllMocks();
    jest.resetModules();
  });

  it('renders the <Li> component', () => {
    const actualOutput = render(<Li>Item</Li>);
    expect(actualOutput).toMatchSnapshot();
  });

  it('renders the <Li> component inside a <Ol>', () => {
    const actualOutput = render(
      <Ol>
        <Li>Item 1</Li>
      </Ol>,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('renders the <Li> component inside a <Ul>', () => {
    const actualOutput = render(
      <Ul>
        <Li>Item 1</Li>
      </Ul>,
    );
    expect(actualOutput).toMatchSnapshot();
  });
});
