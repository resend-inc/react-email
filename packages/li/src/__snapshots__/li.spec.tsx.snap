// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render renders the <Li> component 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><li>Item</li>"`;

exports[`render renders the <Li> component inside a <Ol> 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><ol style=\\"list-style-type:decimal-leading-zero\\"><li>Item 1</li></ol>"`;

exports[`render renders the <Li> component inside a <Ul> 1`] = `"<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\"><ul style=\\"list-style-type:circle\\"><li>Item 1</li></ul>"`;
